### PR TITLE
Chapter 14: Following users

### PR DESCRIPTION
**Pull request by rashidfirhat98**

### CODE REVIEW

**Comment by rashidfirhat98** [comment]
ready

File: app/controllers/relationships_controller.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+class RelationshipsController < ApplicationController
+  before_action :logged_in_user
+
+  def create
+    @user = User.find_by(id: params[:followed_id])
+    current_user.follow(@user)
```
Comment (saobv-2223) [inline app/controllers/relationships_controller.rb]: remove ()

File: app/controllers/relationships_controller.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    respond_to do |format|
+      format.html { redirect_to @user }
+      format.js
+    end
+  end
+
+  def destroy
+    @user = Relationship.find_by(id: params[:id]).followed
+    current_user.unfollow(@user)
```
Comment (saobv-2223) [inline app/controllers/relationships_controller.rb]: remove ()

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -49,6 +49,20 @@ def destroy
     flash[:success] = "User deleted"
     redirect_to users_url
   end
+  
+  def following
+    @title = "Following"
+    @user = User.find_by(id: params[:id])
+    @users = @user.following.page(params[:page])
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]: remove ()

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -49,6 +49,20 @@ def destroy
...
+    render "show_follow"
+  end
+
+  def followers
+    @title = "Followers"
+    @user = User.find_by(id: params[:id])
+    @users = @user.followers.page(params[:page])
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]: remove ()

File: app/helpers/microposts_helper.rb [hunk 5]
```diff
@@ -0,0 +1,2 @@
+module MicropostsHelper
+end
```
Comment (saobv-2223) [inline app/helpers/microposts_helper.rb]: remove file that never used

File: app/views/static_pages/home.html.erb [hunk 6]
```diff
@@ -29,4 +18,20 @@
...
+      sample application.
+    </h2>
+
+    <%= link_to "Sign up now!",  signup_path , class: "btn btn-lg btn-primary" %> 
+  </div>
+
+  <%= link_to image_tag("rails.svg", alt: "Rails logo", width: "200px"), "https://rubyonrails.org/" %>
 <% end %>
+
+
```
Comment (saobv-2223) [inline app/views/static_pages/home.html.erb]:
```suggestion
```

File: app/views/users/_follow_form.html.erb [hunk 7]
```diff
@@ -0,0 +1,9 @@
+<% unless current_user?(@user) %>
```
Comment (saobv-2223) [inline app/views/users/_follow_form.html.erb]: remove ()

File: app/views/users/show.html.erb [hunk 8]
```diff
@@ -19,3 +20,5 @@
     <% end %>
   </div>
 </div>
+
+
```
Comment (saobv-2223) [inline app/views/users/show.html.erb]: remove ()

File: app/views/users/show_follow.html.erb [hunk 9]
```diff
@@ -0,0 +1,32 @@
+<% provide(:title, @title) %>
```
Comment (saobv-2223) [inline app/views/users/show_follow.html.erb]: remove ()

File: app/views/users/show_follow.html.erb [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+  <aside class="col-md-4">
+    <section class="user_info">
+      <%= gravatar_for @user %>
+      <h1><%= @user.name %></h1>
+      <span><%= link_to "view my profile", @user %></span>
+      <span><b>Microposts:</b>
+        <%= @user.microposts.count %></span>
+    </section>
+    <section class="stats">
+      <%= render 'shared/stats' %>
```
Comment (saobv-2223) [inline app/views/users/show_follow.html.erb]: using "

**Comment by rashidfirhat98** [comment]
ready

File: app/controllers/relationships_controller.rb [hunk 11]
```diff
@@ -0,0 +1,21 @@
+class RelationshipsController < ApplicationController
+  before_action :logged_in_user
+
+  def create
+    @user = User.find_by id: params[:followed_id]
```
Comment (vanvtt-0952) [inline app/controllers/relationships_controller.rb]: plz, handling when `@user` is nil 
change by steps
 + remove this line
 + add method into this file
`private`
`def load_user`
` @user = User.find_by id: params[:followed_id]`
`return if @user`
`flash[:warning] = "User not found"`
`redirect_to root_path` 

+ then use callback for action create
add code `before_action :load_user, only: :create`

File: app/controllers/relationships_controller.rb [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+    current_user.follow @user
+    respond_to do |format|
+      format.html { redirect_to @user }
+      format.js
+    end
+  end
+
+  def destroy
+    @user = Relationship.find_by id: params[:id].followed
```
Comment (vanvtt-0952) [inline app/controllers/relationships_controller.rb]: plz create method for `Relationship.find_by id: params[:id]`
and remember handle when the result is nil

Comment (rashidfirhat98) [reply app/controllers/relationships_controller.rb]: is this right? im not sure

**Comment by rashidfirhat98** [comment]
ready

**Review by saobv-2223 — COMMENTED** [review]
merge

**Pull request merged**